### PR TITLE
Add PIC Baseline ASM Plugin

### PR DESCRIPTION
**Pull request by thestr4ng3r**

### CODE REVIEW

File: libr/asm/p/asm_picbaseline.c [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+		break;
+	case PIC_BASELINE_OP_ARGS_8K:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x",
+				  op_info->mnemonic,
+				  instr & 0xff);
+		break;
+	case PIC_BASELINE_OP_ARGS_9K:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x",
+				  op_info->mnemonic,
+				  instr & 0b111111111);
```
Comment (radare) [inline libr/asm/p/asm_picbaseline.c]: 0b is not portable. and this value have 9 1s? so it should be longlong? better use constants with comments or defines to clarify that

File: libr/asm/p/asm_picbaseline.c [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+				  instr & 0b11111);
+		break;
+	case PIC_BASELINE_OP_ARGS_3B_5F:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x, 0x%x",
+				  op_info->mnemonic,
+				  instr & 0b11111,
+				  (instr & (0b111 << 5)) >> 5);
+		break;
+	case PIC_BASELINE_OP_ARGS_8K:
+		snprintf (op->buf_asm, R_ASM_BUFSIZE + 1, "%s 0x%x",
```
Comment (radare) [inline libr/asm/p/asm_picbaseline.c]: this can result on non-null terminated string

File: libr/asm/p/asm_picbaseline.c [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+	}
+}
+
+static int asm_picbaseline_disassemble(RAsm *a, RAsmOp *op, const ut8 *b, int l) {
+#define EMIT_INVALID { \
+	op->size = 1; \
+	strncpy (op->buf_asm, "invalid", R_ASM_BUFSIZE); \
+	return 1; \
+}
+	if (!b || l<2) {
```
Comment (radare) [inline libr/asm/p/asm_picbaseline.c]: missing spaces around <

File: libr/asm/p/asm_picbaseline.c [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+	switch ((instr >> 6) & 0b111111) {
+	case 0b000111:
+		return PIC_BASELINE_OPCODE_ADDWF;
+	case 0b000101:
+		return PIC_BASELINE_OPCODE_ANDWF;
+	case 0b000001:
+		if (instr & 0b100000) {
+			return PIC_BASELINE_OPCODE_CLRF;
+		} else if ((instr & 0x11111) == 0b00000) {
```
Comment (radare) [inline libr/asm/p/asm_picbaseline.c]: no need to else after return

**Comment by Maijin** [comment]
This should be handled by e asm.cpu instead of 2 archs no?

**Comment by thestr4ng3r** [comment]
Maybe it should be merged with pic18c at some point, yes. However I am not 100% sure yet. PIC Baseline and PIC18 both have completely different instruction sets and for both types there are also different cpus with different registers and so on.

**Comment by radare** [comment]
Agree with maijin. We can merge this, but i would prefer to have this as asm.cou variants

**Comment by thestr4ng3r** [comment]
What I meant by merging would be exactly what you meant, to have one plugin called pic with baseline and pic18 cpus. I'll think about how to do the refactoring properly.Am 01.01.2018 21:26 schrieb radare <notifications@github.com>:Agree with maijin. We can merge this, but i would prefer to have this as asm.cou variants

**Comment by radare** [comment]
Why not just name it pic? Picbaseline seems too long to me

**Comment by thestr4ng3r** [comment]
That's what I wrote ;-)Am 01.01.2018 21:58 schrieb radare <notifications@github.com>:Why not just name it pic? Picbaseline seems too long to me

**Comment by XVilka** [comment]
@thestr4ng3r yes, see example e.g. of TMS320 variations handling.

File: libr/asm/p/asm_picbaseline.c (hunk 1 above)
Comment (thestr4ng3r) [reply libr/asm/p/asm_picbaseline.c]: I'll replace the 0b stuff. But why would 9 bits be longlong?

File: libr/asm/p/asm_picbaseline.c (hunk 2 above)
Comment (thestr4ng3r) [reply libr/asm/p/asm_picbaseline.c]: No, snprintf takes the number of bytes, not characters. But it would make sense to replace it with sizeof(...).

**Comment by radare** [comment]
Nvm :D i was drunk or sleepy

**Comment by radare** [comment]
I know. But even sizeof will not enforce null terminated

**Comment by thestr4ng3r** [comment]
@radare Yeah but it's ok the way it is now or am I missing something?
Basically like this:
```
char test[5];
snprintf(test, sizeof(test), "123456");
```
which writes "1234\0".

**Comment by thestr4ng3r** [comment]
Now there is only one plugin for both anal and asm called pic which has baseline and pic18 cpus. I called it pic18 instead of pic18c because pic18c is more specific than it has to be for what the plugin does. (Correct my if I'm wrong, I'm not a PIC expert)

Anal for baseline is not implemented yet, but I plan to do that. The current state can probably be merged for now though.

**Comment by radare** [comment]
great... now r2 masgter doesn tbuilds at all

**Comment by thestr4ng3r** [comment]
Yeah seems like I forgot to commit the plugins cfg files. Fixed anyway now.

**Comment by radare** [comment]
Yep i fixed it. Thanks

**Pull request merged**